### PR TITLE
time-set: adjust system clock if rtc is far in future

### PR DESCRIPTION
**Pull request by Blarse**
Some devices reset the hardware clock to a time far in the future.
If the time of the hardware clock differs from the epoch by more than a threshold value, assume that the hardware clock is wrong.

Also move epoch file to /var

### CODE REVIEW

File: src/shared/clock-util.c [hunk 1]
```diff
@@ -137,12 +137,15 @@ int clock_reset_timewarp(void) {
         return 0;
 }
 
-#define EPOCH_FILE "/usr/lib/clock-epoch"
+#define EPOCH_FILE "/var/lib/systemd/clock-epoch"
```
Comment (bluca) [inline src/shared/clock-util.c]: Why? This is a build-time epoch, it's supposed to be in the read-only vendor subsystem

File: src/core/main.c [hunk 2]
```diff
@@ -1600,9 +1600,9 @@ static void initialize_clock(void) {
 
         r = clock_apply_epoch();
         if (r < 0)
-                log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
+                log_error_errno(r, "Current system time diverges from epoch time, but cannot correct: %m");
```
Comment (bluca) [inline src/core/main.c]: Why this change? The previous version is more accurate

File: src/shared/clock-util.c [hunk 3]
```diff
@@ -137,12 +137,15 @@ int clock_reset_timewarp(void) {
...
+#define DELTA_THRESHOLD (USEC_PER_YEAR*8)
```
Comment (bluca) [inline src/shared/clock-util.c]: this will break long running systems - ie, the support window for the civil infranstructure lts is 15 years

File: src/shared/clock-util.c (hunk 1 above)
Comment (Blarse) [reply src/shared/clock-util.c]: My bad, I was planning to update the timestamp in this file periodically and started thinking of it like changeable file. I'll revert this change.

File: src/core/main.c (hunk 2 above)
Comment (Blarse) [reply src/core/main.c]: Because the system time can be not only **before**, but also far after than the build time(on some systems the rtc reset to the future), and I need to detect this and adjust the system time.

File: src/shared/clock-util.c (hunk 3 above)
Comment (Blarse) [reply src/shared/clock-util.c]: If you think that 15 years is enough, it will be 15.

**Comment by Blarse** [comment]
@bluca, Thanks for taking your time to review my PL.

File: src/shared/clock-util.c (hunk 1 above)
Comment (poettering) [reply src/shared/clock-util.c]: @Blarse timesyncd maintains a dynamic touch file already, that it touches whenever it gets an NTP sync.

this static file here should remain in /usr

File: src/shared/clock-util.c (hunk 3 above)
Comment (poettering) [reply src/shared/clock-util.c]: please make this configurable via meson param. and better pick a value too large than too short.

File: src/core/main.c (hunk 2 above)
Comment (poettering) [reply src/core/main.c]: i don't like this error message. I should clearly say what was tried here, i.e. jump back or forward, given that the error situation is quite a different once.

i.e. change clock_apply_epoch() to return backward or forward jump info somehow. Or call it twice, once for backward and once for forward jumps, with a new param that discerns the two oprations.

File: src/shared/clock-util.c [hunk 4]
```diff
@@ -138,11 +138,14 @@ int clock_reset_timewarp(void) {
 }
 
 #define EPOCH_FILE "/usr/lib/clock-epoch"
+#define DELTA_THRESHOLD (USEC_PER_YEAR*15)
```
Comment (poettering) [inline src/shared/clock-util.c]: we generally suffix defines that refer to time values with `_USEC` to make the unit clear

File: src/shared/clock-util.c [hunk 5]
```diff
@@ -152,7 +155,12 @@ int clock_apply_epoch(void) {
         } else
                 epoch_usec = timespec_load(&st.st_mtim);
 
-        if (now(CLOCK_REALTIME) >= epoch_usec)
+        now_usec = now(CLOCK_REALTIME);
+        diff_usec = usec_sub_unsigned(now_usec, epoch_usec);
+        if(!diff_usec)
```
Comment (poettering) [inline src/shared/clock-util.c]: as per coding style: don't do numeric comparison with C's downgrade-to-bool feature. i.e. do an explicit check == 0 check.

also space after the `if` and before the `(`.

But really, don't use usec_sub_unsigned() here unless you know that the first param is larger than the second. It's ugly to rely on usec_sub_unsigned()'s internal handling of this case. i.e. make this check explicit, by comparing now_usec with epoch_usec first.

File: meson.build [hunk 6]
```diff
@@ -722,6 +722,8 @@ if time_epoch == -1
 endif
 conf.set('TIME_EPOCH', time_epoch)
 
+conf.set('CLOCK_DIFF_THRESHOLD', get_option('clock-diff-threshold'))
```
Comment (poettering) [inline meson.build]: not a fan of needless abbreviations, i.e. "diff" is not pretty.

also, this should make clear in its naming that this is a *maximum* threshold that is applied.

i.e. CLOCK_VALID_RANGE_MAX or so.

File: src/shared/clock-util.c [hunk 7]
```diff
@@ -152,11 +155,22 @@ int clock_apply_epoch(void) {
...
+                ret = 1;
+        else if (diff_usec > (usec_t) CLOCK_DIFF_THRESHOLD * USEC_PER_SEC)
+                ret = 2;
+
+        if (ret == 0)
                 return 0;
 
         if (clock_settime(CLOCK_REALTIME, timespec_store(&ts, epoch_usec)) < 0)
-                return -errno;
+                return -ret;
```
Comment (poettering) [inline src/shared/clock-util.c]: uh, no, please propgate the error code we actually saw, as before. eating up error codes just like that is not ok

File: src/shared/clock-util.c [hunk 8]
```diff
@@ -152,11 +155,22 @@ int clock_apply_epoch(void) {
...
+        now_usec = now(CLOCK_REALTIME);
+        if (now_usec > epoch_usec)
+                diff_usec = usec_sub_unsigned(now_usec, epoch_usec);
+        else
+                diff_usec = usec_sub_unsigned(epoch_usec, now_usec);
+
+        if (now_usec < epoch_usec)
+                ret = 1;
+        else if (diff_usec > (usec_t) CLOCK_DIFF_THRESHOLD * USEC_PER_SEC)
+                ret = 2;
```
Comment (poettering) [inline src/shared/clock-util.c]: nope. no arbitrary constants like this. C has enums for a reason. Also, use the retur value for error codes please, see below.

Add a new output param:

```c
typedef ClockChangeDirection {
        CLOCK_CHANGE_NOOP
        CLOCK_CHANGE_FORWARD,
        CLOCK_CHANGE_BACKWARD,
} ClockChangeDirection;
```

and then change `clock_apply_epoch()` to a protype like this:

```c
int clock_apply_epoch(ClockChangeDirection *attempted_change) {
…
```

And then return the attempted direction change in that.

**Comment by Blarse** [comment]
@poettering Thanks for your patience and willing to work with newbie contributors,
appreciate it, here is the next version.

File: src/shared/clock-util.c [hunk 9]
```diff
@@ -139,10 +139,12 @@ int clock_reset_timewarp(void) {
 
 #define EPOCH_FILE "/usr/lib/clock-epoch"
 
-int clock_apply_epoch(void) {
+int clock_apply_epoch(ClockChangeDirection *ret_attempted_change) {
         struct stat st;
         struct timespec ts;
         usec_t epoch_usec;
+        usec_t now_usec;
+        usec_t diff_usec;
```
Comment (bluca) [inline src/shared/clock-util.c]: You can fold these declarations on the line above

File: src/shared/clock-util.c [hunk 10]
```diff
@@ -139,10 +139,12 @@ int clock_reset_timewarp(void) {
...
```
Comment (bluca) [inline src/shared/clock-util.c:151]: `assert(ret_attempted_change);`

File: src/shared/clock-util.h [hunk 11]
```diff
@@ -3,9 +3,15 @@
 
 #include <time.h>
 
+typedef enum ClockChangeDirection {
+        CLOCK_CHANGE_NOOP,
+        CLOCK_CHANGE_FORWARD,
+        CLOCK_CHANGE_BACKWARD,
```
Comment (bluca) [inline src/shared/clock-util.h:10]:
```
        _CLOCK_CHANGE_MAX,
        _CLOCK_CHANGE_INVALID = -EINVAL,
```

File: src/shared/clock-util.h [hunk 12]
```diff
@@ -2,10 +2,19 @@
 #pragma once
 
 #include <time.h>
+#include <errno.h>
```
Comment (bluca) [inline src/shared/clock-util.h:4]: sort alphabetically

File: src/core/main.c [hunk 13]
```diff
@@ -1598,11 +1598,19 @@ static void initialize_clock(void) {
...
                 (void) clock_reset_timewarp();
 
-        r = clock_apply_epoch();
-        if (r < 0)
-                log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
-        else if (r > 0)
-                log_info("System time before build time, advancing clock.");
+        ClockChangeDirection change_dir;
+        r = clock_apply_epoch(&change_dir);
+        if (r < 0){
```
Comment (bluca) [inline src/core/main.c]: missing space between `)` and `{` (same below)

File: meson_options.txt [hunk 14]
```diff
@@ -208,6 +208,8 @@ option('status-unit-format-default', type : 'combo',
        description : 'use unit name or description in messages by default')
 option('time-epoch', type : 'integer', value : '-1',
        description : 'time epoch for time clients')
+option('clock-valid-range-max', type : 'integer', value : '473364000', # 15 years
+       description : 'maximum value in seconds for the difference between rtc and epoch, exceeding which is considered an rtc error')
```
Comment (poettering) [inline meson_options.txt]: we generally unify on usec time times, please use that here too. and please also include USEC in the name of the setting to indicate the unit: CLOCK_VALID_RANGE_USEC_MAX or so

File: src/shared/clock-util.c [hunk 15]
```diff
@@ -152,7 +154,20 @@ int clock_apply_epoch(void) {
...
+                diff_usec = usec_sub_unsigned(now_usec, epoch_usec);
+        else
+                diff_usec = usec_sub_unsigned(epoch_usec, now_usec);
+
+        *ret_attempted_change = CLOCK_CHANGE_NOOP;
+
+        if (now_usec < epoch_usec)
+                *ret_attempted_change = CLOCK_CHANGE_FORWARD;
+        else if (diff_usec > (usec_t) CLOCK_VALID_RANGE_MAX * USEC_PER_SEC)
+                *ret_attempted_change = CLOCK_CHANGE_BACKWARD;
```
Comment (poettering) [inline src/shared/clock-util.c:164]: why not merge these two into if/else blocks into one:

```c
now_usec = now(CLOCK_REALTIME);
if (now_usec < epoch_usec) 
        *ret_attempted_change = CLOCK_CHANGE_FORWARD;
else if (now_usec > usec_add(epoch_usec, CLOCK_VALID_RANGE_USEC_MAX))
        *ret_attempted_changed = CLOCK_CHANGE_BACKWARD;
else {
        *ret_attempted_change = CLOCK_CHANGE_NOP;
         return 0;
}
```

File: src/shared/clock-util.c [hunk 16]
```diff
@@ -139,10 +139,12 @@ int clock_reset_timewarp(void) {
 
 #define EPOCH_FILE "/usr/lib/clock-epoch"
 
-int clock_apply_epoch(void) {
+int clock_apply_epoch(ClockChangeDirection *ret_attempted_change) {
         struct stat st;
         struct timespec ts;
-        usec_t epoch_usec;
+        usec_t epoch_usec, now_usec, diff_usec;
+
```
Comment (poettering) [inline src/shared/clock-util.c:146]: so usually we have the rule that we do not mangle return params on failures. In this case that's kinda impotant though. Hence please add an explicit comment here, i.e. add something like this here:

```c
/* NB: we update *ret_attempted_change in *all* cases, both on success and failure, to indicate what we intended to do! */
```

File: src/core/main.c [hunk 17]
```diff
@@ -1598,11 +1598,19 @@ static void initialize_clock(void) {
...
+        if (r < 0) {
+                if (change_dir == CLOCK_CHANGE_FORWARD)
+                        log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
+                else if (change_dir == CLOCK_CHANGE_BACKWARD)
+                        log_error_errno(r, "Current system time is far ahead build time, but cannot correct: %m");
+        } else if (r > 0) {
+                if (change_dir == CLOCK_CHANGE_FORWARD)
+                        log_info("System time before build time, advancing clock.");
+                else if (change_dir == CLOCK_CHANGE_BACKWARD)
+                        log_info("System time far ahead build time, reversing clock.");
```
Comment (poettering) [inline src/core/main.c]: I think proper english would be "far ahead of".

maybe even add a "too" in there. But maybe mention the extra time we give here.

```
log_info("System time further ahead than %s after build time, resetting clock to build time.", FORMAT_TIMESPAN(CLOCK_VALID_RANGE_USEC_MAX, USEC_PER_DAY));
```

File: src/shared/clock-util.c [hunk 18]
```diff
@@ -152,8 +157,20 @@ int clock_apply_epoch(void) {
...
-        if (now(CLOCK_REALTIME) >= epoch_usec)
+        now_usec = now(CLOCK_REALTIME);
+        if (now_usec > epoch_usec)
+                diff_usec = usec_sub_unsigned(now_usec, epoch_usec);
+        else
+                diff_usec = usec_sub_unsigned(epoch_usec, now_usec);
+
+        if (now_usec < epoch_usec)
+                *ret_attempted_change = CLOCK_CHANGE_FORWARD;
+        else if (diff_usec > (usec_t) CLOCK_VALID_RANGE_USEC_MAX)
```
Comment (poettering) [inline src/shared/clock-util.c]: my comment on this was not addressed?

File: src/core/main.c [hunk 19]
```diff
@@ -1598,11 +1599,20 @@ static void initialize_clock(void) {
...
                 (void) clock_reset_timewarp();
 
-        r = clock_apply_epoch();
-        if (r < 0)
-                log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
-        else if (r > 0)
+        ClockChangeDirection change_dir;
+        r = clock_apply_epoch(&change_dir);
+        if (r == 0)
+                log_info("Initialized system clock");
```
Comment (bluca) [inline src/core/main.c]: nit: missing period. Also, maybe make this `log_debug`? Finally, the system clock is changed only if `r == 1`, no? In case `r == 0` it's a no-op?

Comment (bluca) [reply src/core/main.c]: so maybe just remove this message altogether, as it would be confusing (says clock was initialized, but nothing was actually done)

Comment (Blarse) [reply src/core/main.c]: done

File: src/core/main.c [hunk 20]
```diff
@@ -1598,11 +1599,18 @@ static void initialize_clock(void) {
                  */
                 (void) clock_reset_timewarp();
 
-        r = clock_apply_epoch();
-        if (r < 0)
-                log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
-        else if (r > 0)
+        ClockChangeDirection change_dir;
+        r = clock_apply_epoch(&change_dir);
+        if(r > 0 && change_dir == CLOCK_CHANGE_FORWARD)
```
Comment (bluca) [inline src/core/main.c]: style nit: missing space after `if`

Comment (bluca) [reply src/core/main.c]: (same below)

Comment (Blarse) [reply src/core/main.c]: oops, thanks

**Comment by poettering** [comment]
lgtm

**Review by bluca — APPROVED** [review]
lgtm too
